wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+				);
+			case 'simpleExpr':
+				return $this->translate_simple_expr( $ast );
+			case 'predicateOperations':
+				$token = $ast->get_child_token();
+				if ( WP_MySQL_Lexer::LIKE_SYMBOL === $token->id ) {
+					return $this->translate_like( $ast );
+				} elseif ( WP_MySQL_Lexer::REGEXP_SYMBOL === $token->id ) {
+					return $this->translate_regexp_functions( $ast );
+				}
+				return $this->translate_sequence( $ast->get_children() );
+			case 'runtimeFunctionCall':
+				return $this->translate_runtime_function_call( $ast );
+			case 'functionCall':
+				return $this->translate_function_call( $ast );
+			case 'systemVariable':
+				// @TODO: Emulate some system variables, or use reasonable defaults.
+				//        See: https://dev.mysql.com/doc/refman/8.4/en/server-system-variable-reference.html
+				//        See: https://dev.mysql.com/doc/refman/8.4/en/server-system-variables.html
+
+				// When we have no value, it's reasonable to use NULL.
+				return 'NULL';
+			case 'castType':
+				// Translate "CAST(... AS BINARY)" to "CAST(... AS BLOB)".
+				if ( $ast->has_child_token( WP_MySQL_Lexer::BINARY_SYMBOL ) ) {
+					return 'BLOB';
+				}
+				return $this->translate_sequence( $ast->get_children() );
+			case 'defaultCollation':
+				// @TODO: Check and save in information schema.
+				return null;
+			case 'duplicateAsQueryExpression':
+				// @TODO: How to handle IGNORE/REPLACE?
+
+				// The "AS" keyword is optional in MySQL, but required in SQLite.
+				return 'AS ' . $this->translate( $ast->get_child_node() );
+			case 'indexHint':
+			case 'indexHintList':
+				return null;
+			default:
+				return $this->translate_sequence( $ast->get_children() );
+		}
+	}
+
+	private function translate_token( WP_MySQL_Token $token ) {
+		switch ( $token->id ) {
+			case WP_MySQL_Lexer::EOF:
+				return null;
+			case WP_MySQL_Lexer::AUTO_INCREMENT_SYMBOL:
+				return 'AUTOINCREMENT';
+			case WP_MySQL_Lexer::BINARY_SYMBOL:
+				/*
+				 * There is no "BINARY expr" equivalent in SQLite. We can look for
+				 * the BINARY keyword in particular cases (with REGEXP, LIKE, etc.)
+				 * and then remove it from the translated output here.
+				 */
+				return null;
+			case WP_MySQL_Lexer::SQL_CALC_FOUND_ROWS_SYMBOL:
+				/*
+				 * The "SQL_CALC_FOUND_ROWS" keyword is implemented in the select
+				 * statement translation and then removed from the output here.
+				 */
+				return null;
+			default:
+				return $token->value;
+		}
+	}
+
+	private function translate_sequence( array $nodes, string $separator = ' ' ): ?string {
+		$parts = array();
+		foreach ( $nodes as $node ) {
+			if ( null === $node ) {
+				continue;
+			}
+
+			$translated = $this->translate( $node );
+			if ( null === $translated ) {
+				continue;
+			}
+			$parts[] = $translated;
+		}
+		if ( 0 === count( $parts ) ) {
+			return null;
+		}
+		return implode( $separator, $parts );
+	}
+
+	private function translate_string_literal( WP_Parser_Node $node ): string {
```

Comment:
cc @dmsnell – there's a lot of escaping happening here, and I know you have deep thoughts and knowledge about MySQL escaping

Reply:
@JanJakes this is the part where things may go very wrong, let's make sure we have a large and comprehensive test suite just for this

Reply:
@adamziel As mentioned in https://github.com/Automattic/sqlite-database-integration/pull/1#discussion_r1946737335, I'll create a ticket for this, will add a larger test case, and we can discuss further there.